### PR TITLE
feat: support upgrades

### PR DESCRIPTION
**Pull request by mkenigs**
Add an `upgrades` field to `Environment` that can be used to upgrade specific packages or all packages.

### CODE REVIEW

**Comment by mkenigs** [comment]
This is ready for review but waiting on https://github.com/flox/pkgdb/pull/142

**Review by garbas — APPROVED** [review]

File: src/resolver/environment.cc [hunk 1]
```diff
@@ -129,6 +129,25 @@ Environment::groupIsLocked( const InstallDescriptors & group,
 
   for ( auto & [iid, descriptor] : group )
     {
+      /* Check for upgrades. */
```
Comment (aakropotkin) [inline src/resolver/environment.cc:132]: non-blocking educational note:
 
For `std::variant` a `std::get_if` like this is totally fine, no need to change it.
A _preferred_ approach is using a `switch { case: ... }` style construct which ensures that you've _really_ checked all variant types.
```c++
class Empty {};
std::variant<int, Empty> value = 1;
std::visit( overloaded {
  []( int x ) { std::cout << "Integer: " << x << std::endl; },
  []( Empty e ) { std::cout << "Empty" << std::endl; }
}, value );
```

The definition of `overloaded` will trigger compilation warnings/errors if you've skipped any variant types.

File: src/resolver/environment.cc [hunk 2]
```diff
@@ -129,6 +129,25 @@ Environment::groupIsLocked( const InstallDescriptors & group,
...
+      if ( bool * upgradeEverything = std::get_if<bool>( &this->upgrades ) )
+        {
+          /* If we're upgrading everything, the group needs to be locked again.
+           */
```
Comment (aakropotkin) [inline src/resolver/environment.cc:136]:
```suggestion
          /* If we're upgrading everything, the group needs to be locked again. */
```

File: src/resolver/environment.cc [hunk 3]
```diff
@@ -129,6 +129,25 @@ Environment::groupIsLocked( const InstallDescriptors & group,
...
+          if ( *upgradeEverything ) { return false; }
+        }
+      else
+        {
+          /* If the current iid is being upgraded, the group needs to be locked
+           * again.
+           */
```
Comment (aakropotkin) [inline src/resolver/environment.cc:143]:
```suggestion
          /* If the current iid is being upgraded, the group needs to be
           * locked again. */
```

I'm just nick picking, don't mind me <3

**Review by aakropotkin — APPROVED** [review]
Minor changes indicated in patches, otherwise LGTM

**Pull request merged**